components/patient-data/migrations/src/main/java/org/phenotips/data/internal/R54691PhenoTips1378DataMigration.java

```diff
@@ -0,0 +1,143 @@
+ *
+ * @version $Id$
+ * @since 1.1M1
+ */
+@Component
+@Named("R54691PhenoTips#1378")
+@Singleton
+public class R54691PhenoTips1378DataMigration extends AbstractHibernateDataMigration
+{
+    /** Logging helper object. */
+    @Inject
+    private Logger logger;
+
+    /** Resolves unprefixed document names to the current wiki. */
+    @Inject
+    @Named("current")
+    private DocumentReferenceResolver<String> resolver;
+
+    @Override
+    public String getDescription()
+    {
+        return "Adding a default 'private' visibility object to all patients which did not have a "
+            + "visibility object #1378";
+    }
+
+    @Override
+    public XWikiDBVersion getVersion()
+    {
+        return new XWikiDBVersion(54691);
+    }
+
+    @Override
+    public void hibernateMigrate() throws DataMigrationException, XWikiException
+    {
+        getStore().executeWrite(getXWikiContext(), new AddVisObjectCallback());
+    }
+
+    /**
+     * Searches for all documents containing vcf files and replaces the reference genome name with the GRCh notation.
+     */
+    private final class AddVisObjectCallback implements XWikiHibernateBaseStore.HibernateCallback<Object>
+    {
+        @Override
+        public Object doInHibernate(Session session) throws HibernateException, XWikiException
+        {
+
+            XWikiContext context = R54691PhenoTips1378DataMigration.this.getXWikiContext();
+            XWiki xwiki = context.getWiki();
+            DocumentReference classReference =
+                new DocumentReference(context.getDatabase(), "PhenoTips", "VisibilityClass");
+            Query q1 =
+                session
+                    .createQuery("select doc.fullName from XWikiDocument as doc, BaseObject "
+                    + "as patobj where doc.fullName=patobj.name "
+                        + "and patobj.className='PhenoTips.PatientClass'");
+            Query q2 =
+                session
+                    .createQuery("select doc.fullName from XWikiDocument as doc, BaseObject patobj, BaseObject visobj "
+                    + "where doc.fullName=patobj.name and "
+                        + "patobj.className='PhenoTips.PatientClass' and "
+                        + "visobj.className='PhenoTips.VisibilityClass' and doc.fullName=visobj.name");
+            @SuppressWarnings("unchecked")
+            Set<String> documents = new HashSet<String>(q1.list());
+            documents.removeAll(q2.list());
+            R54691PhenoTips1378DataMigration.this.logger.debug("Found {} documents with no visibility object",
+                documents.size());
+            for (String docName : documents) {
+                R54691PhenoTips1378DataMigration.this.logger.debug("Checking [{}]", docName);
+                XWikiDocument doc =
+                    xwiki.getDocument(R54691PhenoTips1378DataMigration.this.resolver.resolve(docName), context);
+                BaseObject visobj = new BaseObject();
```

Comment:
Because it already does all that's needed to make sure the object is correctly tied to the document and has the right xclass. See all that needs to be done to have that information correctly set up [here](https://github.com/xwiki/xwiki-platform/blob/xwiki-platform-5.4.6/xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java#L2077).